[v6r14] Added CheckCECapabilities pilot command

CinziaLu opened the pull request:
- Modified Resource.py utility in order to get list of all the Tags from ce/queue getQueue() to avoid that during the merging of the ceDict and queueDict the duplicate keys are overwritten.

- Added the dirac-resource-get-parameters script that retrieves all the
parameters of a given Site. To be used as following:
  
  dirac-resource-get-parameters -S site_name -C ce_name -Q queueName
  
- Updated the pilotCommand with the CheckCECapabilities command which
uses the script above to add the CE/Queue tags, if any, to the dirac.cfg

atsareg commented on WorkloadManagementSystem/PilotAgent/pilotCommands.py:
```diff
@@ -359,6 +358,42 @@ def _getSecurityCFG( self ):
+
+  def execute( self ):
+
+    retCode, resourceDict = self.executeAndGetOutput( 'dirac-resource-get-parameters -S %s -N %s -Q %s' %
```
I may be miss something but why can not you use the Resources helper getQueue() here ? The DIRAC client is already installaed and configured by this point - isn't it ?

fstagni replied:
Installed, configured, but at no point the parseCommandLine() is done... but in the script

atsareg replied:
I wonder if just gConfig.forceRefresh() will do the trick ?

fstagni replied:
So, after test: we can't use gConfig here, and this is clear as soon as we do "from DIRAC import gConfig" or any other "from DIRAC* import *" in any part of the pilot code. 2 reasons:
- if (VO)DIRAC is installed with dirac-install, it will still complain that can't find GSI. And this is because the environment at this point is not set (it's not in the current process)
- if (VO)DIRAC is installed from, say, CVMFS, still the environment created has to be passed in to the subprocess that run the scripts.

atsareg replied:
  Hi Federico, Cinzia,
      
        I see, thanks for having tested it. So, let's continue with the
      current implementation by calling the
      external command. We have to still think how the usual DIRAC
      modules can be used once it is
      installed. This is not trivial indeed but can be quite beneficial.
      
        Cheers,
        Andrei 
      
      On 26/10/15 13:00, fstagni wrote:
    
    
      In WorkloadManagementSystem/PilotAgent/pilotCommands.py:
      > @@ -359,6 +358,42 @@ def _getSecurityCFG( self ):
>        self.cfg.append( "-o /DIRAC/Security/CertFile=%s/hostcert.pem" % self.pp.certsLocation )
>        self.cfg.append( "-o /DIRAC/Security/KeyFile=%s/hostkey.pem" % self.pp.certsLocation )
>  
> +class CheckCECapabilities( CommandBase ):
> +  """ Used to get  CE tags.
> +  """
> +  def __init__( self, pilotParams ):
> +    """ c'tor
> +    """
> +    super( CheckCECapabilities, self ).__init__( pilotParams )
> +    self.cfg = []
> +
> +
> +  def execute( self ):
> +
> +    retCode, resourceDict = self.executeAndGetOutput( 'dirac-resource-get-parameters -S %s -N %s -Q %s' %

      So, after test: we can't use gConfig here, and this is clear as
        soon as we do "from DIRAC import gConfig" or any other "from
        DIRAC* import *" in any part of the pilot code. 2 reasons:
      
        if (VO)DIRAC is installed with dirac-install, it will still
          complain that can't find GSI. And this is because the
          environment at this point is not set (it's not in the current
          process)
        if (VO)DIRAC is installed from, say, CVMFS, still the
          environment created has to be passed in to the subprocess that
          run the scripts.
      
      —
        Reply to this email directly or view
          it on GitHub.

CinziaLu commented:
I used the external script for the CheckCECapabilities.
I added the CheckWNCapabilities pilot command to retrieve processors and memory. 
For the memory I added the JobMemory module which provide getMemoryFromMJF() if mjf is available and the getMemoryFromProc() to get the memory from proc/meminfo. Here we have to agree if this is the memory parameter we are looking for. 
I added the dirac-wms-get-wn-parameters script which uses the JobMemory and the Os.getNumberOfCores() to retreive processors and memory.

atsareg commented:
Is it tested enough to go to the v6r14 patch ?

fstagni commented:
I tested myself the pilot and it's alright. When it's merged and tagged I can anyway have a better test using our Jenkins tests, that I can also update with what's specific of this PR.

I didn't test the SiteDirector.

fstagni commented:
Anyway, when this is merged, we need to advertise the changes because extensions (including LHCb) might need to include these commands in the list of executed commands, if there is a custom list.

CinziaLu commented:
I tested the getQueues() of the SiteDirector. We agreed with Federico that the manipulation of the tags fields are not necessary anymore. I let them be a string, as they are coming from the resourceDict. According to this, the memoryTags are added to the string of tags instead of to a list of tags.  However, they could also be removed if we will use the new CheckWNCapabilities command.

atsareg commented on WorkloadManagementSystem/Agent/SiteDirector.py:
```diff
@@ -237,25 +234,16 @@ def getQueues( self, resourceDict ):
             if memoryTags:
               self.queueDict[queueName]['ParametersDict'].setdefault( 'Tag', [] )
-              self.queueDict[queueName]['ParametersDict']['Tag'] += memoryTags
+              self.queueDict[queueName]['ParametersDict']['Tag'] += ','.join( memoryTags )
```
Why did you change the Tag parameter evaluation for the queue ? This parameter is needed fully when checking for available jobs for this queue. Including CE tags.

atsareg replied:
Also the Tag parameter in the CE description used in the job matching is supposed to be a list and not a comma separated string

CinziaLu replied:
Because, we agreed with Federico, that this step will be done by the CheckCECapabilities command, where the Resource.getQueues() is doing it. Otherwise we would do the step twice, in siteDirector and in CheckCECapabilities. I let Federico comment on that.

atsareg replied:
This is a misunderstanding. What we agreed about is that SiteDirector will not pass the Tags as an argument of the pilot. The tags are needed for the SiteDirector logic because they are used in the evaluation of availability of payload before submitting pilots to a particular CE/queue. So, I suggest that you remove the SiteDirector changes from this PR and I will add the suppression of the Tag pilot argument in my PR for the SiteDirector ( different problem ) to avoid conflicts

fstagni replied:
I have a question here: the moment the SiteDirector is selecting if there are pilots to send, do we really need that TAGs become matching parameters? Because at this moment I would live them out.

atsareg replied:
Why ? Tags are resource description parameters as others. Imagine that you have jobs in the TQ all requiring a certain software. Then there is no point to send pilots to sites where this software is not installed.

fstagni replied:
Well, this can become a longer discussion if we should or we shouldn't. 
So, I would say that for today, since this is a patch, let's agree on 
your proposal of 2 comments above.

atsareg commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -135,19 +135,34 @@ def getStorageElementOptions( seName ):
+  if ceTags:
+    if isinstance( ceTags, basestring ):
+      Tags = fromChar( ceTags )
+    else:
```
I think this is a redundant if/else: . With getOptionsDict you get all the options as strings by definition.

atsareg commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -135,19 +135,34 @@ def getStorageElementOptions( seName ):
   resultDict.update( result['Value'] )
+  queueTags = resultDict.get( 'Tag' )
+  if queueTags:
+    if isinstance( queueTags , basestring ):
```
Same as above

atsareg commented:
I merge it now. But still one comment. JobMemory is not a client, this is rather what we call Utilities. So, I suggest to move it to WorkloadManagementSystem/Utilities folder. Client are classes that encapsulate communications with remote services ( mostly ). Those classes can be extended eventually in the DIRAC extensions. It is important to keep the software architecture clean.

fstagni commented:
I just re-iterate that: when this is merged, we need to advertise the changes because extensions (including LHCb) might need to include these commands in the list of executed commands, if there is a custom list. So this has to go at least in https://github.com/DIRACGrid/DIRAC/wiki/DIRAC-v6r14

CinziaLu commented:
Moved JobMemory.py

atsareg commented on WorkloadManagementSystem/PilotAgent/pilotCommands.py:
```diff
@@ -359,6 +358,94 @@ def _getSecurityCFG( self ):
+    if not Processors:
+      self.log.warn( "Could not retrieve number of processors" )
+    else:
+      self.pp.tags.append( 'Processors_%s' % Processors )
```
The WN capacity must be communicated in the resource description differently. Add MaxRAM and NumberOfProcessors option in the CEDefaults, and not tags.

CinziaLu replied:
Hi Andrei, I will modify it according to your comment. Following this JIRA task https://its.cern.ch/jira/browse/LHCBDIRAC-468 I was using tags. I would really appreciate to discuss better the tasks to be sure we agree on what it is needed to be done.

fstagni commented on WorkloadManagementSystem/Utilities/JobMemory.py:
```diff
@@ -0,0 +1,37 @@
+
+def getMemoryFromProc():
+    meminfo = dict( ( i.split()[0].rstrip( ':' ), int( i.split()[1] ) ) for i in open( '/proc/meminfo' ).readlines() )
+    mem_free_kb = meminfo['MemFree']
```
This has to be in MB

fstagni replied:
And also let's put "MemTotal"

fstagni commented on WorkloadManagementSystem/Utilities/JobMemory.py:
```diff
@@ -0,0 +1,37 @@
+
+def getMemoryFromProc():
+    meminfo = dict( ( i.split()[0].rstrip( ':' ), int( i.split()[1] ) ) for i in open( '/proc/meminfo' ).readlines() )
+    MemTotal = meminfo['MemFree']
```
No, I meant to use
```
meminfo['MemTotal']
```
instead of 
```
meminfo['MemFree']
```

fstagni commented on WorkloadManagementSystem/scripts/dirac-wms-get-wn-parameters.py:
```diff
@@ -0,0 +1,23 @@
+"""
+
+from DIRAC.Core.Base import Script
+from DIRAC import gConfig, gLogger, exit as DIRACExit
```
This is not used

fstagni commented on WorkloadManagementSystem/scripts/dirac-wms-get-wn-parameters.py:
```diff
@@ -0,0 +1,23 @@
+"""
+  Determine number of processors and memory for the worker node
+"""
+
```
Missing __RCSID__

fstagni commented:
These new scripts are missing the shebang.

fstagni commented:
Review OK
